### DISCUSSION TITLE
Results on the Cloud Detection dataset

### DISCUSSION THREAD

**Discussion by robmarkcole** [torchgeo/torchgeo#3184] [Show and tell]
I thought I would document my efforts using the new `CloudCoverDetectionDataModule` for the Sentinel 2 [Cloud cover detection](https://torchgeo.readthedocs.io/en/latest/api/datasets.html#cloud-cover-detection) dataset. I would be interested to hear if anyone can improve on them. 

I will start with my concerns about the data quality - many of the annotations don't appear accurate, sample below:

<img width="477" height="483" alt="image" src="https://github.com/user-attachments/assets/6a71794b-5a86-4850-bc62-9416f61be6a1" />

<img width="667" height="679" alt="image" src="https://github.com/user-attachments/assets/0224ce76-4384-4f2e-adaf-1d238c279c53" />

Then there is a class imbalance, it is weighted with more clear pixels than cloudy, but is similar across the train/test splits:
```
⚡ ~ python scripts/cloudcover_label_balance.py --root data/cloudcover/public/train_labels
Scanned 11748 label files under data/cloudcover/public/train_labels
Total valid pixels: 3,079,667,712
value 0: 1,165,210,450 pixels (37.8356 %)
value 1: 1,914,457,262 pixels (62.1644 %)

⚡ ~ python scripts/cloudcover_label_balance.py --root data/cloudcover/private/test_labels
Scanned 10980 label files under data/cloudcover/private/test_labels
Total valid pixels: 2,878,341,120
value 0: 1,117,741,256 pixels (38.8328 %)
value 1: 1,760,599,864 pixels (61.1672 %)
```

I wrapped the datamodule to provide some additional augmentations: 
```python

class CloudCoverDetectionAugmentedDataModule(CloudCoverDetectionDataModule):
    """CloudCoverDetection DataModule with augmentations (no resize)."""

    def __init__(
        self,
        batch_size: int = 12,
        num_workers: int = 0,
        val_split_pct: float = 0.2,
        **kwargs,
    ):
        super().__init__(
            batch_size=batch_size,
            num_workers=num_workers,
            val_split_pct=val_split_pct,
            **kwargs,
        )

        # --- Augmentations borrowed from VHR10DataModule (minus resize) ---
        self.train_aug = K.AugmentationSequential(
            K.Normalize(mean=self.mean, std=self.std),
            K.RandomHorizontalFlip(),
            K.RandomVerticalFlip(),
            data_keys=None,
            keepdim=True,
        )

        self.aug = K.AugmentationSequential(
            K.Normalize(mean=self.mean, std=self.std),
            data_keys=None,
            keepdim=True,
        )
```

The wining entries using this dataset [used](https://drivendata.co/blog/cloud-cover-winners/) `focal` loss, although I didn't see a big benefit from using it. I settled on:
```python
batch_size = 16

datamodule = CloudCoverDetectionAugmentedDataModule(
    root='/teamspace/studios/this_studio/data/cloudcover',
    download=False,
    batch_size=batch_size,
)

task = SemanticSegmentationTask(
    model="unet",
    backbone="resnet18",
    weights=True,
    in_channels=4,
    num_classes=2,
    # loss="focal",
    labels=['background', 'cloud'],
    class_weights=Tensor([1.64, 1.0]),
    lr=1e-4,
)
``` 
I trained for 3-10 epochs in various experiments on an L4 - it was pretty slow going so I didn't train longer. Train loss is pretty bumpy - may benefit from a larger batch size on a bigger machine:

<img width="621" height="285" alt="image" src="https://github.com/user-attachments/assets/24f04baa-a6a7-4665-b992-791df8bd55c9" />

Val loss trending in the right direction:
<img width="621" height="280" alt="image" src="https://github.com/user-attachments/assets/bfbad792-c631-46a4-96fe-dfbe1a2dbaac" />

Final metrics:
```
[{'test_loss': 302.2618713378906,
  'test_multiclassaccuracy_background': 0.1536271721124649,
  'test_multiclassaccuracy_cloud': 0.9917892217636108,
  'test_AverageAccuracy': 0.5727081894874573,
  'test_AverageF1Score': 0.5256727933883667,

  'test_AverageJaccardIndex': 0.400911808013916,

  'test_AveragePrecision': 0.7872213125228882,
  'test_AverageRecall': 0.5727081894874573,
  'test_multiclassfbetascore_background': 0.26331833004951477,
  'test_multiclassfbetascore_cloud': 0.7880272269248962,
  'test_multiclassjaccardindex_background': 0.15162153542041779,
  'test_multiclassjaccardindex_cloud': 0.6502020955085754,
  'test_OverallAccuracy': 0.670783281326294,
  'test_OverallF1Score': 0.670783281326294,
  'test_OverallJaccardIndex': 0.5046455264091492,
  'test_OverallPrecision': 0.670783281326294,
  'test_OverallRecall': 0.670783281326294,
  'test_multiclassprecision_background': 0.9207216501235962,
  'test_multiclassprecision_cloud': 0.6537209749221802,
  'test_multiclassrecall_background': 0.1536271721124649,
  'test_multiclassrecall_cloud': 0.9917892217636108}]
```
Some predictions:
<img width="827" height="556" alt="image" src="https://github.com/user-attachments/assets/58bfef25-fc75-478f-8d4c-f00f10100b6a" />

<img width="821" height="556" alt="image" src="https://github.com/user-attachments/assets/c79242fc-89ca-48a6-9080-b2188abb2fb4" />

**Comment by robmarkcole** [comment]
Suggestions on slack:

- try using a shallow network to create pseudo labels (unable to overfit to noisy labels), then train with those.
-  try ranges of label smoothing
- U-Net efficient-net backbone is really good

More encouraging result using `jaccard` loss, stopped after 5 epochs:
```
{'test_loss': 0.5401086807250977,
  'test_multiclassaccuracy_background': 0.26139506697654724,
  'test_multiclassaccuracy_cloud': 0.9771360158920288,
  'test_AverageAccuracy': 0.6192655563354492,
  'test_AverageF1Score': 0.6025365591049194,

  'test_AverageJaccardIndex': 0.46104276180267334,

  'test_AveragePrecision': 0.7785675525665283,
  'test_AverageRecall': 0.6192655563354492,
  'test_multiclassfbetascore_background': 0.4026944935321808,
  'test_multiclassfbetascore_cloud': 0.8023785948753357,
  'test_multiclassjaccardindex_background': 0.252108633518219,
  'test_multiclassjaccardindex_cloud': 0.6699768900871277,
  'test_OverallAccuracy': 0.7030158638954163,
  'test_OverallF1Score': 0.7030158638954163,
  'test_OverallJaccardIndex': 0.5420389175415039,
  'test_OverallPrecision': 0.7030158638954163,
  'test_OverallRecall': 0.7030158638954163,
  'test_multiclassprecision_background': 0.8764879107475281,
  'test_multiclassprecision_cloud': 0.6806471943855286,
  'test_multiclassrecall_background': 0.26139506697654724,
  'test_multiclassrecall_cloud': 0.9771360158920288}]
```

<img width="962" height="312" alt="image" src="https://github.com/user-attachments/assets/050157ac-57de-4d51-bfec-a18669b37104" />

<img width="963" height="312" alt="image" src="https://github.com/user-attachments/assets/29581a8d-a923-42bd-9d0d-1279160f9784" />

<img width="961" height="309" alt="image" src="https://github.com/user-attachments/assets/dfa432cb-b6cd-44d4-8978-09260357e05e" />

**Comment by robmarkcole** [comment]
Try more epochs, we are targetting Jaccard index (test_AverageJaccardIndex) of at least 0.89

## 10 Epochs
```
[{'test_loss': 0.44047728180885315,
  'test_multiclassaccuracy_background': 0.4362524151802063,
  'test_multiclassaccuracy_cloud': 0.9622360467910767,
  'test_AverageAccuracy': 0.6992442607879639,
  'test_AverageF1Score': 0.7075627446174622,

  'test_AverageJaccardIndex': 0.5620198249816895,

  'test_AveragePrecision': 0.8054649829864502,
  'test_AverageRecall': 0.6992442607879639,
  'test_multiclassfbetascore_background': 0.5827997326850891,
  'test_multiclassfbetascore_cloud': 0.8323257565498352,
  'test_multiclassjaccardindex_background': 0.411233127117157,
  'test_multiclassjaccardindex_cloud': 0.7128065228462219,
  'test_OverallAccuracy': 0.7607906460762024,
  'test_OverallF1Score': 0.7607906460762024,
  'test_OverallJaccardIndex': 0.6139322519302368,
  'test_OverallPrecision': 0.7607906460762024,
  'test_OverallRecall': 0.7607906460762024,
  'test_multiclassprecision_background': 0.8776089549064636,
  'test_multiclassprecision_cloud': 0.7333210706710815,
  'test_multiclassrecall_background': 0.4362524151802063,
  'test_multiclassrecall_cloud': 0.9622360467910767}]
```

## 25 epochs
Overfitting now?
```
[{'test_loss': 0.600573718547821,
  'test_multiclassaccuracy_background': 0.16034744679927826,
  'test_multiclassaccuracy_cloud': 0.9801923036575317,
  'test_AverageAccuracy': 0.5702698826789856,
  'test_AverageF1Score': 0.5263516902923584,

  'test_AverageJaccardIndex': 0.3998751640319824,

  'test_AveragePrecision': 0.7434409856796265,
  'test_AverageRecall': 0.5702698826789856,
  'test_multiclassfbetascore_background': 0.268981009721756,
  'test_multiclassfbetascore_cloud': 0.7837224006652832,
  'test_multiclassjaccardindex_background': 0.15538880228996277,
  'test_multiclassjaccardindex_cloud': 0.6443614959716797,
  'test_OverallAccuracy': 0.6662015914916992,
  'test_OverallF1Score': 0.6662015914916992,
  'test_OverallJaccardIndex': 0.4994770288467407,
  'test_OverallPrecision': 0.6662015914916992,
  'test_OverallRecall': 0.6662015914916992,
  'test_multiclassprecision_background': 0.8340194821357727,
  'test_multiclassprecision_cloud': 0.652862548828125,
  'test_multiclassrecall_background': 0.16034744679927826,
  'test_multiclassrecall_cloud': 0.9801923036575317}]
```
<img width="614" height="282" alt="image" src="https://github.com/user-attachments/assets/2649b5e5-d7eb-474a-a29f-1bf3121d8ed2" />

<img width="608" height="265" alt="image" src="https://github.com/user-attachments/assets/966772c0-b63d-4b6d-b621-bf042dbac16b" />

<img width="731" height="494" alt="image" src="https://github.com/user-attachments/assets/8a330756-3cd8-42dd-91ac-84cce119f5ef" />
